### ISSUE TITLE
When computing the perturbation, do we need to set model.eval()? 

### ISSUE THREAD

**Issue by vtddggg** [locuslab/fast_adversarial#14]
Hello Leslie Rice and Eric Wong,

Congratulations on your significant work！！

I found the model is always set to `training` mode during adversarial training period. However, I think when we compute the adversarial perturbation, we must set `model.eval()` to prevent the randomness, such as dropout, to affect the estimated gradients. So a correct approach is to add `model.eval()` before [this line](https://github.com/locuslab/fast_adversarial/blob/54f728755e71857b632882ba6d6cef22f56e2172/CIFAR10/train_fgsm.py#L112).

I'm curious about why you did not set `model.eval()` in your code. I guess the `amp` makes the gradient overflow in `eval` mode? How about the performance gap between these two different approaches?

Looking forward to your reply, thank you !

**Comment by cmhzc** [comment]
I found that not setting `model.eval()` actually enhances robust accuracy in this case

**Comment by leslierice1** [comment]
You are correct, typically the model is set to eval mode while performing the attack. I don't believe we omitted this step intentionally, however when I re-run the default CIFAR-10 configuration, the resulting model gets slightly worse PGD test accuracy (44.34% as opposed to 44.90%).
